design/proposals/modular-components.md

```diff
@@ -0,0 +1,234 @@
+   of the _ksonnet-lib_ and _kubernetes_ versions being used by the respective
+   environment.
+4. There is a `base.libsonnet` file which acts as the execution entry point
+   when interacting with the clusters specified by a environment. Each
+   environment also contains a `main.libsonnet` file which acts as the
+   "`base-libsonnet` override" file for each environment; allowing for
+   functionality such as the exclusion of components from the specific
+   environment.
+5. Each environment also contains a `params.libsonnet` file that allows for
+   tweaking of simple component parameters (_i.e._ MySQL host) for each
+   environment.
+
+## Proposal
+
+1. Retire `main.jsonnet` and `base.libsonnet`.
+   * The purpose of these files are often confusing to users. The goal of
+   `main.jsonnet` was to act as a way for environments to provide
+   environment-specific extensibility to components. However, this only works
+   if the component is written in jsonnet, and also requires the user to have
+   an advanced understanding of jsonnet.
+   * Furthermore, with the existing model, all components are operated on
+   during `apply`, `delete`, etc. If the components vary drastically between
+   environments, it is not convenient to "subtract" these components from each
+   environment's `main.jsonnet`.
+   * `base.libsonnet` also currently serves no purpose besides acting as an entry
+   point that _imports_ all components.
+
+2. Consolidate environment metadata files.
+   * Each environment's `.metadata` directory currently store 3 files related to
+   the generated `ksonnet-lib`. This is unecessary and also costly as the
+   number of environments grow. We needn't store multiple copies of the same
+   API version on disk. The Kubernetes API version that each environment uses
+   can be recorded in the environment specification file. The metadata files
+   can be locally cached once per k8s API version in `lib`.
+
+3. A component should not be uniquely defined by it's base file name.
+   * The existing implementation implies that there cannot be a `foo.jsonnet`
+   and a `foo.yaml`, since the base file name is shared.
+
+4. Handle environment specifications in `app.yaml` and retire `spec.json`.
+   * Multiple files and areas for configuration can cause confusions for users.
+   In order to keep configurations simple, I propose that all ksonnet-specific
+   _app-level_ configurations reside in the `app.yaml`.
+
+`app.yaml` would add an _environments_ section as follows:
+
+```
+apiVersion: 0.0.1
+kind: ksonnet.io/app
+name: config
+version: 0.0.1
+registries:
+  incubator:
+    gitVersion:
+      commitSha: ea3408d44c2d8ea4d321364e5533d5c60e74bce0
+      refSpec: master
+    protocol: github
+    uri: github.com/ksonnet/parts/tree/master/incubator
+environments:
+  dev:
+    k8sVersion: 1.9.0
+    server: https://my-dev-server.com
+    namespaces:
+      - dev-foo
+      - dev-bar
+    targets:
```

Comment:
IMHO more often than not, every environment would have same amount of targets. Making users to specify all targets for all envs feels like optimizing for less common case and would cause errors such as new component was specified in `dev`, tested and verifies, then app goes to `prod` and component is missing and app is on fire.

Reply:
I disagree, this is something we would heavily use where I work.

Feature flagging is a common thing I do today with helm to disable specific parts of the stack being deployed depending on the environment. It's very common to build and add a new microservice to the stack, but it's often not deployed to all environments until it's been fleshed out in dev more.

I think having an explicit list of targets is also hugely important for avoiding accidental changes when a component changes but the `app.yaml` doesn't.  This pattern also exists in many other places like config management, where you really want to be explicit about what's being ran for each environment, in case the upstream adds a ton of new components, and you want to ensure you're only deploying what's needed when updating it.

Reply:
The real world use case we've seen is when dev and production differ because of security requirements. The goal here is to give a good default and allow for more flexibility when it is needed.

Reply:
Also, I believe this could be accommodated by defining a component `default.jsonnet` which just effectively imports the other components, and in the `app.yaml`, you would just specify `targets: [ "default.jsonnet" ]`.

Reply:
@chancez ,

> It's very common to build and add a new microservice to the stack, but it's often not deployed to all environments until it's been fleshed out in dev more.

Looks like you are using single ksonnet app to describe whole stack, something which would be pulling in tens of components, which vary between environment depending how mature they are. if so is it the right scope for single ksonnet app?  One of the promises of microservices is independent release cycle, instead of one big bang, by pulling them all under single umbrella you deny that kind of flexibility.

Maybe there is a different view of what component is? For me ksonnet app is a single app which provides a single service. It might require some dependencies for it's own operation, but these dependencies are more like implementation detail of that app. Dependencies which are shared in implementation (but not during operation) make sense to factor out as  components, for example: redis, memcache, rabbitmq, etc.   From this POV, there will be very rare change between environments regarding components app is installing: if it requires redis for its operation, it is going to need it in every environment.

Reply:
@chancez ,

> I think having an explicit list of targets is also hugely important for avoiding accidental changes when a component changes but the app.yaml doesn't.

I can't understand which failure scenario you are describing, would you mind to elaborate? Components in ksonnet so far are installed prototypes, which themselve are  small pieces of code calling package they come from in a particular way.  Package versions are in full control of user, they can't accidentally change. All prototypes from a package need be installed explicitly and become component to have any effect on ksonnet app. Simple fact that there is  a component is already explicit enough to avoid surprises.

Reply:
I think both ways makes sense. I believe that ideally, you can build tons of individually deployable components, and that should work. But I think being able to deploy them together via ksonnet is also important. Whether you use ksonnet to deploy groups of apps, or apps individually depends on size of teams and various other factors.

Just because you can develop and release microservices individually for example, doesn't mean that's how it will work in all scenarios. You might be deploying and releasing microservices individually in pre-prod, but it's entirely likely for production things are released together.

Reply:
> , but it's entirely likely for production things are released together.

Of course you can choose to release them together, but should it really be done as a single ksonnet app describing whole stack? There are simply no tooling in ksonnet to orchestrate such whole stack rollout as a single app: no dependencies, no inbetween smoke tests, no custom success conditions and notifications.

Reply:
I'd argue that's a problem of ksonnet already, today.

Ex, if you're deploying $microService and Redis together, you're already in a situation where you cannot orchestrate the rollout nor the interaction between these two components ,and they're coupled to each other. In dev, who cares, in staging/prod, maybe they should each be deployed individually with things like smoke tests running between each component deployment.

I'm not saying grouping of many resources in a single deployment is the best practice, but it's a very reasonable one depending on your needs. Many systems or applications simply don't need to do anything other than roll out all new things together. I entirely agree if you need to run various tests, and other things between rollouts, you'll likely need tooling on top of ksonnet, and your components will need to be broken out accordingly, but I think having flexibility here is important. 

Having to orchestrate many individual deploys itself is complex, and often it doesn't need to be. People are already fine running `kubectl apply -f` or `helm install` on large groups of resources, without a heavy need for handling each individual component's roll out.

Reply:
There is a big difference, that components of an app are all managed by the same team, where upgrade path is excersised much more often and there are much less communication distance than between teams. Orchestrating update of a whole stack, especially when it done for the first time in prod (as dev is done independently as per discussion above) is  a much biggger task and it is not purely technical. There is also difference in scope: single digit number of compoonents can be coordinated with initcontainers and entrypoints if needed, but more often than not just putting them all there is enough.

Anyway, this thread is not about how companies should manage their CD pipeline, it is about not optimizing for case which ksonnet is less than an optimal tool right now.  I am all for explicit over implicit, but I am also against unnecessary duplication: simple fact that component is there is explicit enough to indicate that ksonnet app needs it.

What about defaulting list of targets to all components  and using explicit `targets:` if specified only? This way spcial environments can list all their components, but more common cases will be covered automatically: all present components are installed.

Reply:
I think i'd rather see a feature where could expose different lists of targets. Perhaps a concept like a role. You could require a default role be defined, which would contain the list of targets to use by default if targets isn't explicitly set. This would also then allow different lists of targets to be defined, and used. Of course, you could set targets that list components directly also.

I can definitely see in the long run if the option to specify targets exists (which I think we can agree is useful) then how people write their apps will change. Potentially writing many components, but not expecting all of them to be installed. Having a way to declare the default list of them, and alternatives would be much more flexible. 

In the end, you should be able to define what components to expose for the default installation, and not necessarily all should be used by default.

Reply:
What I failed to mention in the proposal is if a target isn't specified, all components will be targeted by default. This is parallel to how ksonnet operates currently, so it should solve both scenarios mentioned in this conversation.